File: DataInjectorAndroid/DataInjectorLib/src/main/java/com/crescentflare/datainjector/injector/FilterNullInjector.java
package com.crescentflare.datainjector.injector;

import com.crescentflare.datainjector.utility.InjectorUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Data injector: remove null values
 * Traverses a data set recursively removing null values
 */
public class FilterNullInjector extends DataInjector
{
    // ---
    // Initialization
    // ---

    public FilterNullInjector()
    {
    }


    // ---
    // Injection
    // ---

    public void apply(Object targetData, Object referencedData, Object subReferencedData)
    {
        Map<String, Object> targetMap = InjectorUtil.asStringObjectMap(targetData);
        List<Object> targetList = InjectorUtil.asObjectList(targetData);
        if (targetMap != null)
        {
            processMap(targetMap);
        }
        else if (targetList != null)
        {
            processList(targetList);
        }
    }


    // ---
    // Dependencies
    // ---

    public List<String> findDependencies()
    {
        return new ArrayList<>();
    }


    // ---
    // Helper
    // ---

    private void processList(List<Object> array)
    {
        List<Object> modifiedList = new ArrayList<>();
        for (Object arrayItem : array)
        {
            Map<String, Object> targetMap = InjectorUtil.asStringObjectMap(arrayItem);
            List<Object> targetList = InjectorUtil.asObjectList(arrayItem);
            if (targetMap != null)
            {
                processMap(targetMap);
                modifiedList.add(targetMap);
            }
            else if (targetList != null)
            {
                processList(targetList);
                modifiedList.add(targetList);
            }
            else if (arrayItem != null)
            {
                modifiedList.add(arrayItem);
            }
        }
        array.clear();
        array.addAll(modifiedList);
    }

    private void processMap(Map<String, Object> map)
    {
        List<String> keys = new ArrayList<>();
        for (String key : map.keySet())
        {
            keys.add(key);
        }
        for (String key : keys)
        {
            Object value = map.get(key);
            Map<String, Object> mapValue = InjectorUtil.asStringObjectMap(value);
            List<Object> listValue = InjectorUtil.asObjectList(value);
            if (mapValue != null)
            {
                processMap(mapValue);
            }
            else if (listValue != null)
            {
                processList(listValue);
            }
            else if (value == null)
            {
                map.remove(key);
            }
        }
    }
}

File: DataInjectorIOS/DataInjector/Classes/injector/FilterNullInjector.swift
//
//  FilterNullInjector.swift
//  DataInjector Pod
//
//  Library injector: remove null values
//  Traverses a data set recursively removing null values
//

import Foundation

open class FilterNullInjector: DataInjector {
    
    // ---
    // MARK: Initialization
    // ---
    
    public override init() {
    }

    
    // ---
    // MARK: Injection
    // ---

    override open func appliedInjection(targetData: Any, subTargetData: Any?, referencedData: Any? = nil, subReferencedData: Any? = nil) -> Any {
        if let dictItem = targetData as? [String: Any] {
            return processDict(dict: dictItem)
        } else if let arrayItem = targetData as? [Any] {
            return processArray(array: arrayItem)
        }
        return targetData
    }
    

    // ---
    // MARK: Dependencies
    // ---

    override open func foundDependencies() -> [String] {
        return []
    }
    

    // ---
    // MARK: Helper
    // ---
    
    private func processArray(array: [Any]) -> [Any] {
        var modifiedArray: [Any] = []
        for i in 0..<array.count {
            if let dictItem = array[i] as? [String: Any] {
                modifiedArray.append(processDict(dict: dictItem))
            } else if let arrayItem = array[i] as? [Any] {
                modifiedArray.append(processArray(array: arrayItem))
            } else if !(array[i] is NSNull) {
                modifiedArray.append(array[i])
            }
        }
        return modifiedArray
    }
    
    private func processDict(dict: [String: Any]) -> [String: Any] {
        var modifiedDict: [String: Any] = [:]
        for (key, value) in dict {
            if let dictItem = value as? [String: Any] {
                modifiedDict[key] = processDict(dict: dictItem)
            } else if let arrayItem = value as? [Any] {
                modifiedDict[key] = processArray(array: arrayItem)
            } else if !(value is NSNull) {
                modifiedDict[key] = value
            }
        }
        return modifiedDict
    }
    
}
